tools/ontology-builder/src/all_ontology_generator.py

```diff
@@ -110,43 +110,50 @@ def _load_ontology_object(onto_file: str) -> owlready2.entity.ThingClass:
+    of distance from the given term.
 
     :param owlready2.entity.ThingClass onto_class: the class for which ancestors will be retrieved
     :param str onto_name: only ancestors from this ontology will be kept
 
     :rtype List[str]
     :return list of ancestors (term ids), it could be empty
     """
-
-    def _get_branch_ancestors(term: owlready2.entity.ThingClass) -> Set[str]:
-        # a branch ancestor is defined in ontology files as having a "part_of" (BFO_0000050) relationship with a term
-        branch_ancestors = set()
-        for subclass in term.is_a:
+    ancestors: Dict[str, int] = dict()
+    queue = [(onto_class, 1)]
+    while queue:
+        term, distance = queue.pop(0)
+        for parent in term.is_a:
+            # a branch ancestor is defined as having a "part_of" (BFO_0000050) relationship with a term
             if (
-                hasattr(subclass, "property")
-                and subclass.property.name == "BFO_0000050"
-                and isinstance(subclass.value, owlready2.entity.ThingClass)
+                hasattr(parent, "property")
+                and parent.property.name == "BFO_0000050"
+                and isinstance(parent.value, owlready2.entity.ThingClass)
             ):
-                branch_ancestor_name = subclass.value.name.replace("obo.", "").replace("_", ":")
-                if branch_ancestor_name.split(":")[0] == onto_name:
-                    branch_ancestors.add(branch_ancestor_name)
-                    branch_ancestors.update(_get_ancestors(subclass.value, onto_name))
-        return branch_ancestors
-
-    ancestors = set()
-    for ancestor in onto_class.ancestors():
-        if ancestor.name.split("_")[0] == onto_name:
-            ancestors.add(ancestor.name.replace("_", ":"))
-            if onto_name != "NCBITaxon":
-                ancestors.update(_get_branch_ancestors(ancestor))
-
-    # remove self
-    ancestors.remove(onto_class.name.replace("_", ":"))
-    return ancestors
+                branch_ancestor_name = parent.value.name.replace("obo.", "").replace("_", ":")
+                if branch_ancestor_name in ancestors:
+                    ancestors[branch_ancestor_name] = min(ancestors[branch_ancestor_name], distance)
+                else:
+                    queue.append((parent.value, distance + 1))
+                    if branch_ancestor_name.split(":")[0] == onto_name:
+                        ancestors[branch_ancestor_name] = distance
+            elif hasattr(parent, "name") and not hasattr(parent, "Classes"):
+                parent_name = parent.name.replace("_", ":")
+                if parent_name in ancestors:
+                    ancestors[parent_name] = min(ancestors[parent_name], distance)
+                else:
+                    queue.append((parent, distance + 1))
+                    ancestors[parent_name] = distance
+
+    # filter out ancestors that are not from the ontology we are currently processing
```

Comment:
It's doable but adds some complexity. The issue is, we still want to traverse the ancestors for non-ontology terms because, as it turns out, some non-ontology terms have ancestors that are a part of the ontology. For instance, the [root node of EFO](https://www.ebi.ac.uk/ols4/ontologies/efo/classes/http%253A%252F%252Fwww.ebi.ac.uk%252Fefo%252FEFO_0000001?lang=en) has 5 direct children, none of which are EFO terms. But, they all have EFO children. So, we still need to traverse these 5 non-EFO children to get to the root node and resolve the ontology graph. But we don't want to count them as ancestors. We also want to make sure we don't traverse them again after they were already visited.

We could add a second "visited_but_not_counted" set and updating it in the for loop, rather than filtering out of the ancestor dict. Whichever you prefer. More performant but more complicated loop logic to read

Reply:
let go for readability in the builder since it is a run once kinda deal.

Reply:
no changes needed